### PR TITLE
Test newly supported interfaces in  Quarkus Spring Data JPA and getReferencedBy as replacement for deprecated methods

### PR DESCRIPTION
**Pull request by michalvavrik**
### Summary

This PR consist of TD for https://github.com/quarkus-qe/quarkus-test-plans/blob/main/QUARKUS-4958.md.

Please select the relevant options.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Dependency update
- [ ] Refactoring
- [ ] Backport
- [x] New scenario (non-breaking change which adds functionality)
- [ ] This change requires a documentation update
- [x] This change requires execution against OCP (use `run tests` phrase in comment)

### Checklist:
- [x] Methods and classes used in PR scenarios are meaningful
- [x] Commits are well encapsulated and follow [the best practices](https://cbea.ms/git-commit/)

### CODE REVIEW

**Comment by michalvavrik** [comment]
run tests

**Comment by michalvavrik** [comment]
@jedla97 I have tried to run OCP locally as there seems to be something wrong with our Jenkins jobs. The OCP CI is passing when run from my workstation.

**Comment by jedla97** [comment]
run tests

**Comment by jedla97** [comment]
@michalvavrik I manually fix it, we yesterday merge jenkins changes, and this seems to be error which I miss. I'll will be providing fix later.

Also look at this PR later today

**Review by jedla97 — APPROVED** [review]
Thanks! PR look great and I don't have any comment's to actual code.

I have just question. You wanted test `TypedSort` but I don't see it anywhere. What I read it is/should be used when sorting by multiple arguments. 
Also with this test multiple sorting parameters come to my mind. I don't know if it's supported usecase in Quarkus but Spring data should support it. It should be possible like `?sort=name,asc&sort=age,desc`.

**Comment by michalvavrik** [comment]
> I have just question. You wanted test TypedSort but I don't see it anywhere. What I read it is/should be used when sorting by multiple arguments. Spring data should support it.

I have tried `TypedSort in `quarkus/extensions/spring-data-jpa/deployment` test `BookPagingAndSortingRepositoryTest.shouldReturnAllBooksSortedByNameDescResult` and when I typed `var typedSort = Sort.TypedSort.sort(Book.class).by((Function<Book, Object>) Book::getName);` I got

```
java.lang.NoClassDefFoundError: org/springframework/aop/framework/ObjenesisCglibAopProxy
	at org.springframework.aop.framework.DefaultAopProxyFactory.createAopProxy(DefaultAopProxyFactory.java:71)
	at org.springframework.aop.framework.ProxyCreatorSupport.createAopProxy(ProxyCreatorSupport.java:105)
	at org.springframework.aop.framework.ProxyFactory.getProxy(ProxyFactory.java:110)
	at org.springframework.data.util.MethodInvocationRecorder.create(MethodInvocationRecorder.java:100)
	at org.springframework.data.util.MethodInvocationRecorder.forProxyOf(MethodInvocationRecorder.java:76)
	at org.springframework.data.domain.Sort$TypedSort.<init>(Sort.java:687)
	at org.springframework.data.domain.Sort.sort(Sort.java:143)
	at io.quarkus.spring.data.deployment.BookPagingAndSortingRepositoryTest.shouldReturnAllBooksSortedByNameDescResult(BookPagingAndSortingRepositoryTest.java:95)
	at io.quarkus.spring.data.deployment.BookPagingAndSortingRepositoryTest_Subclass.shouldReturnAllBooksSortedByNameDescResult$$superforward(Unknown Source)
	at io.quarkus.spring.data.deployment.BookPagingAndSortingRepositoryTest_Subclass$$function$$2.apply(Unknown Source)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:73)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:62)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.invokeInOurTx(TransactionalInterceptorBase.java:136)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.invokeInOurTx(TransactionalInterceptorBase.java:107)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired.doIntercept(TransactionalInterceptorRequired.java:38)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.intercept(TransactionalInterceptorBase.java:61)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired.intercept(TransactionalInterceptorRequired.java:32)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired_Bean.intercept(Unknown Source)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:42)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:30)
	at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:27)
	at io.quarkus.spring.data.deployment.BookPagingAndSortingRepositoryTest_Subclass.shouldReturnAllBooksSortedByNameDescResult(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at io.quarkus.test.QuarkusUnitTest.runExtensionMethod(QuarkusUnitTest.java:513)
	at io.quarkus.test.QuarkusUnitTest.interceptTestMethod(QuarkusUnitTest.java:427)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: java.lang.ClassNotFoundException: org.springframework.aop.framework.ObjenesisCglibAopProxy
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:526)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:542)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:490)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:542)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:490)
	... 27 more
```
and more I think about it, neither https://quarkus.io/guides/spring-data-jpa#what-is-supported or https://quarkus.io/guides/spring-data-rest#what-is-supported mentions `TypedSort`. I also checked Quarkus and there is not a single test on `TypedSort`. Finally, `org.springframework.data.domain.Sort#sort` is only way I found to create `TypedSort`, but it invokes always this constructor that causes above-mentioned ^^^. And the same constructor is completely deleted in native substitution: https://github.com/quarkusio/quarkus/blob/main/extensions/spring-data-jpa/runtime/src/main/java/io/quarkus/spring/data/runtime/Target_org_springframework_data_domain_Sort_TypedSort.java

So I think I was wrong to say that I am going to test `TypedSort` as it is not explicitly marked as supported.

> It should be possible like ?sort=name,asc&sort=age,desc.

Yeah, but you see, the difference between `ListPagingAndSortingRepository` and `PagingAndSortingRepository` as far as Quarkus implementation goes is `findAll` method. Quarkus transforms it to same Panache method https://github.com/quarkusio/quarkus/blob/main/extensions/spring-data-jpa/deployment/src/main/java/io/quarkus/spring/data/deployment/generate/StockMethodsAdder.java#L590 and then, in case it is _just_ PagingAndSortingRepository that returns iterator, then it uses this `findAll` that returns list and adds another `declaration` that leverages fact that `List` extends `Iterator`. It's the same result.

TL;DR; I didn't feel the urge to test this when original test coverage didn't. But I'll add extra column in a moment. it shouldn't take me more than few minutes.

**Review by jedla97 — APPROVED** [review]
> TL;DR; I didn't feel the urge to test this when original test coverage didn't. But I'll add extra column in a moment. it shouldn't take me more than few minutes.

Thanks for adding it!

> So I think I was wrong to say that I am going to test TypedSort as it is not explicitly marked as supported.
When i think about it maybe it was just added for spring extension be able to build to native. But thanks for investigate it.

Anyway updated PR looks good to me so if the CI pass we can merge it! There is no need to run OCP as it was pass before and this addition shouldn't break it.

**Comment by michalvavrik** [comment]
Thanks. GH CI is green, so let me merge it so that I can send status and call it a week.
